docs/deep-dive/substances.md

```diff
@@ -0,0 +1,88 @@
+
+Let's imagine that you have a gemstone lab. In this lab, a sample might be a small bit of a larger gemstone.
+
+You would have to tell the LIMS that you're working with gemstones, so you would implement the GemstoneSample type in your plugin:
+
+```
+# TODO: This part of the API is still being worked on. One can currently only
+# interact with the proxied objects mentioned below
+
+class GemstoneSample(Sample):
+    color = TextField()
+    comment = TextField()
+```
+
+The GemstoneSample object is in fact a proxy for a range of database models:
+
+    * ExtensibleType
+    * ExtensiblePropertyType
+    * Substance
+    * ExtensibleProperty
+
+When the system registers your plugin, it finds your `GemstoneSample`. Based on it, the system creates an entry in the table `ExtensibleType`, which registers your `GemstoneSample` class. It also creates two entries in the `ExtensiblePropertyType` table, for each of the properties.
+
+Now you can create an instance of the `Substance` like this:
+
+```
+s1 = GemstoneSample()
+s1.color = "red"
+s1.comment = "some comment"
+s1.save()
+```
+
+Under the hood, this will type check against the `ExtensibleType` and the `ExtensiblePropertyType`s and then save a new `Substance` entry with two new `ExtensibleProperty` entries.
+
+This is achieved by means of the `SubstanceService` class. When you call `s1.save()` above, it will call the service object which in turn will save all of the required objects.
+
+# Working with substances on a low level
+
+NOTE: This section is intended for those that need more control over the creation extensible objects.
+
+When creating a Substance, you'll first need a `ExtensibleType` entry. You can create that with a call to the substances service.
+
+```
+properties = properties or [
+    dict(name='color', raw_type=ExtensiblePropertyType.STRING, display_name='Colour'),
+    dict(name='comment', raw_type=ExtensiblePropertyType.STRING, display_name='Comment'),
+]
+
+substance_type = substances.register_type('GemstoneSample', 'substances', plugin, properties=properties)
+```
+
+We will now have one `ExtensibleType` entry in the DB as well as two `ExtensiblePropertyType` entries.
+
+The next step is to create the `Substance`. One can do this via the `substances` singleton like this:
+
+```
+props = {
+    'comment': 'found this yesterday',
+    'color': 'red'
+}
+
+substance = substances.create(
+    name=name,
+    extensible_type=substance_type,
+    organization=org,
+    properties=properties
+)
+```
+
+This will create one `Substance` where:
+
+    * It is marked as version 1
+    * It has the properties `comment` and `color` set
+    * It has been type checked against the substance type we created before
+    * It has no origins or parents as it's not derived from any sample
+
```

Comment:
I miss an overview showing 1) version of substance, 2) version of property, 3) depth och substance, 4) parent of substance, 5) origin of substance.

Reply:
This document is close related to the sample.md document. Maybe we should put them more close together? Rename it to substance overview, then substance deep dive?

Reply:
Yes, I'll address the first point. I think I'll do another runthrough later to address the second point.

Reply:
After giving it some more thought, I think I'll address these in a later commit as it's still somewhat a moving target :)